tokio-util/src/io/mem.rs

```diff
@@ -0,0 +1,177 @@
+//! IO over in-process memory
+
+use std::{task::{self, Poll, Waker}, pin::Pin, sync::{Arc, Mutex}};
+use bytes::{Buf, BytesMut};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+/// A bidirectional pipe to read and write bytes in memory.
+///
+/// A pair of `MemStream`s are created together, and they act as a "channel"
+/// that can be used as in-memory IO types. Writing to one of the pairs will
+/// allow that data to be read from the other, and vice versa.
+///
+/// # Example
+///
+/// ```
+/// # async fn ex() -> std::io::Result<()> {
+/// # use tokio::io::{AsyncReadExt, AsyncWriteExt};
+/// use tokio_util::io::MemStream;
+///
+/// let (mut client, mut server) = MemStream::pair();
+///
+/// client.write_all(b"ping").await?;
+///
+/// let mut buf = [0u8; 4];
+/// server.read_exact(&mut buf).await?;
+/// assert_eq!(&buf, b"ping");
+///
+/// server.write_all(b"pong").await?;
+///
+/// client.read_exact(&mut buf).await?;
+/// assert_eq!(&buf, b"pong");
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Debug)]
+pub struct MemStream {
+    read: Arc<Mutex<Pipe>>,
+    write: Arc<Mutex<Pipe>>,
+}
+
+/// A unidirectional IO over a piece of memory.
+///
+/// Data can be written to the pipe, and reading will return that data.
+#[derive(Debug)]
+struct Pipe {
+    /// The buffer storing the bytes written, also read from.
+    ///
+    /// Using a `BytesMut` because it has efficient `Buf` and `BufMut`
+    /// functionality already. Additionally, it can try to copy data in the
+    /// same buffer if there read index has advanced far enough.
+    buffer: BytesMut,
+    /// Determines if the write side has been closed.
+    is_closed: bool,
+    /// If the `read` side has been polled and is pending, this is the waker
+    /// for that parked task.
+    read_waker: Option<Waker>,
+}
+
+// ===== impl MemStream =====
+
+impl MemStream {
+    /// Create a new pair of `MemStream`s that act like a pair of connected sockets.
+    pub fn pair() -> (MemStream, MemStream) {
```

Comment:
Would suggest having a buffer size limit; it's often useful to test the behavior of some IO-related code under write backpressure.

Reply:
I'm not sure that's a blocker for this PR, but I broadly agree. I think that behavior could added via a builder that has options along the lines of `.with_inflight_limit()`, `.with_rate_limit`, `.with_drop_rate`, etc. All of these could be driven by some Rng, enabling some simulation-like testing.

Reply:
I thought adding a buffer limit would be useful, but didn't want to figure out that API until after...